File: extensions/extensions-transform/extensions-transform-spark-java/src/main/java/com/boozallen/data/transform/spark/DataTransformer.java
package com.boozallen.data.transform.spark;

/*-
 * #%L
 * aiSSEMBLE::Extensions::Transform::Spark::Java
 * %%
 * Copyright (C) 2021 Booz Allen
 * %%
 * This software package is licensed under the Booz Allen Public License. All Rights Reserved.
 * #L%
 */

import java.util.List;

import com.boozallen.data.transform.DataTransformException;
import org.apache.commons.collections.CollectionUtils;
import org.apache.commons.lang3.StringUtils;
import org.apache.spark.sql.Dataset;
import org.technologybrewery.mash.MediationContext;
import org.technologybrewery.mash.Mediator;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.boozallen.aissemble.core.policy.configuration.configuredrule.ConfiguredRule;
import com.boozallen.data.transform.policy.DataTransformPolicy;
import com.boozallen.data.transform.policy.DataTransformPolicyManager;

/**
 * {@link DataTransformer} class represents the rule engine for looking up a
 * specified data transform policy, invoking the corresponding transformation
 * rules, and then processing the results from rule invocation.
 * 
 * @author Booz Allen Hamilton
 * 
 */
public class DataTransformer {

    private static final Logger logger = LoggerFactory.getLogger(DataTransformer.class);
    private static final DataTransformPolicyManager policyManager = DataTransformPolicyManager.getInstance();

    /**
     * Applies a data transform policy on a dataset.
     * 
     * @param policyIdentifier
     *            the identifier of the policy to apply
     * @param dataset
     *            the dataset to apply the policy on
     * @return the dataset after applying the policy
     */
    @SuppressWarnings("rawtypes")
    public Dataset transform(String policyIdentifier, Dataset dataset) {
        DataTransformPolicy policy = (DataTransformPolicy) policyManager.getPolicy(policyIdentifier);
        List<ConfiguredRule> rules = policy != null ? policy.getRules() : null;

        if (policy == null) {
            throw new DataTransformException(
                    "No data transform policy found for identifier '" + policyIdentifier + "'");
        }

        if (CollectionUtils.isEmpty(rules)) {
            throw new DataTransformException("No rules found in data transform policy '" + policyIdentifier + "'");
        }

        if (rules.size() > 1) {
            throw new DataTransformException("Data transform policy '" + policyIdentifier
                    + "' contains multiple rules, which is currently not supported");
        }

        return invokeMediator(policy, rules.get(0), dataset);
    }

    @SuppressWarnings("rawtypes")
    private Dataset invokeMediator(DataTransformPolicy policy, ConfiguredRule rule, Dataset dataset) {
        String inputType = (String) rule.getConfiguration("inputType");
        String outputType = (String) rule.getConfiguration("outputType");

        if (StringUtils.isBlank(inputType) || StringUtils.isBlank(outputType)) {
            throw new DataTransformException(
                    "Rule inputType and/or outputType configuration is missing in data transform policy '"
                            + policy.getIdentifier() + "'");
        }

        // look up the mediator to invoke
        MediationContext mediationContext = new MediationContext(inputType, outputType);
        Mediator mediator = policy.getMediationManager().getMediator(mediationContext);

        logger.debug("Invoking mediator {}", mediator.getClass().getName());

        return (Dataset) mediator.mediate(dataset);
    }

}

File: extensions/extensions-transform/aissemble-extensions-transform-spark-python/src/data_transform_spark/data_transformer.py
# #%L
# aiSSEMBLE::Extensions::Transform::Spark::Python
# %%
# Copyright (C) 2021 Booz Allen
# %%
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# #L%
###
from typing import Dict
from krausening.logging import LogManager
from pyspark.sql.dataframe import DataFrame
from policy_manager.policy import ConfiguredRule
from data_transform_core import DataTransformException
from data_transform_core.mediator import MediationContext
from data_transform_core.policy import DataTransformPolicyManager, DataTransformPolicy


class DataTransformer:
    """
    DataTransformer class represents the rule engine for looking up a
    specified data transform policy, invoking the corresponding transformation
    rules, and then processing the results from rule invocation.
    """

    __logger = LogManager.get_instance().get_logger("DataTransformer")

    def __init__(self) -> None:
        self._policyManager = DataTransformPolicyManager.getInstance()

    def transform(self, policyIdentifier: str, dataset: DataFrame) -> DataFrame:
        """
        Applies a data transform policy on a dataset.
        """
        policy = self._policyManager.getPolicy(policyIdentifier)
        rules = policy.rules if policy else None

        if not policy:
            raise DataTransformException(
                'No data transform policy found for identifier "%s"' % policyIdentifier
            )

        if not rules:
            raise DataTransformException(
                'No rules found in data transform policy "%s"' % policyIdentifier
            )

        if len(rules) > 1:
            raise DataTransformException(
                'Data transform policy "%s" contains multiple rules, which is currently not supported'
                % policyIdentifier
            )

        return self.invokeMediator(policy, rules[0], dataset)

    def invokeMediator(
        self, policy: DataTransformPolicy, rule: ConfiguredRule, dataset: DataFrame
    ) -> DataFrame:
        inputType = rule.configurations["inputType"]
        outputType = rule.configurations["outputType"]

        if not inputType or not outputType:
            raise DataTransformException(
                'Rule inputType and/or outputType configuration is missing in data transform policy "%s"'
                % policy.identifier
            )

        # Look up the mediator to invoke
        mediationContext = MediationContext(inputType=inputType, outputType=outputType)
        mediator = policy.mediationManager.getMediator(mediationContext)

        DataTransformer.__logger.debug("Invoking mediator %s" % type(mediator))

        return mediator.mediate(dataset)

    def validate(self, input: any, properties: Dict[str, str]) -> None:
        """
        Validates the input and properties for this transformer.
        """
        if not isinstance(input, DataFrame):
            raise DataTransformException("Input (%s) is not a PySpark dataframe")
